### PR TITLE
Introduce RowExpressionRewriter

### PR DESCRIPTION
**Pull request by highker**

### CODE REVIEW

**Comment by highker** [comment]
@rongrong, I kinda wanna enforce the callers (mostly opt rules) to do the check and pick the one to use. So that when I refactor those rules, I can remove those branches with some examination on the logic.

File: presto-main/src/main/java/com/facebook/presto/sql/relational/RowExpressionTreeRewriter.java [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+            return context;
+        }
+
+        public boolean isDefaultRewrite()
+        {
+            return defaultRewrite;
+        }
+    }
+
+    private static <T> boolean sameElements(Optional<T> a, Optional<T> b)
```
Comment (rongrong) [inline presto-main/src/main/java/com/facebook/presto/sql/relational/RowExpressionTreeRewriter.java]: This method doesn't seem to be used?

File: presto-main/src/main/java/com/facebook/presto/sql/relational/RowExpressionNodeInliner.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+public class RowExpressionNodeInliner
+        extends RowExpressionRewriter<Void>
+{
+    public static RowExpression replaceExpression(RowExpression expression, Map<? extends RowExpression, ? extends RowExpression> mappings)
+    {
+        return RowExpressionTreeRewriter.rewriteWith(new RowExpressionNodeInliner(mappings), expression);
+    }
+
+    private final Map<? extends RowExpression, ? extends RowExpression> mappings;
```
Comment (rongrong) [inline presto-main/src/main/java/com/facebook/presto/sql/relational/RowExpressionNodeInliner.java]: Can we move the variables to the top? It took me a while to find it. 😂

File: presto-main/src/test/java/com/facebook/presto/sql/planner/TestRowExpressionRewriter.java [hunk 3]
```diff
@@ -34,17 +36,24 @@
...
 import static com.facebook.presto.sql.relational.Expressions.constant;
+import static com.facebook.presto.sql.relational.RowExpressionNodeInliner.replaceExpression;
 import static org.testng.Assert.assertEquals;
 import static org.testng.Assert.assertTrue;
 
 public class TestRowExpressionRewriter
 {
-    @Test
-    public void testSimple()
+    private FunctionManager functionManager;
```
Comment (rongrong) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/TestRowExpressionRewriter.java]: Would look a little cleaner if you move this logic to first commit. But I don't have strong feelings about this.

**Pull request merged**